### PR TITLE
docs: create REST API endpoints in components + LB3 migration

### PR DESCRIPTION
**Pull request by bajtos**
_This is a part of documentation for the spike #4099 How to migrate LB3 components._

In this pull request, I am covering migration of REST API endpoints.

- Create a new page for component authors to explain how to contribute REST API endpoints
- Create a new page for component authors to explain how to contribute local service classes or providers
- Create a migration guide explaining how to migrate REST API endpoints from LoopBack 3 components to LoopBack 4 style.
This is a part of a spike, so I should not be changing the real docs, but then I figured out that since I have already wrote the content, it will give more value to our users to put the content up, even if it's not finished yet. My intention is to land this pull request even if the content is not in the shape we would eventually want and create follow-up stories to make any substantial improvements as required by reviewers. (This the the part where I am getting back to our usual spike process - do the research and then open follow-up stories to implement the proposed changes.)

@raymondfeng @jannyHou @emonddr: please review

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- Affected artifact templates in `packages/cli` were updated
- Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: docs/site/extending/rest-api.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+endpoints contributed by a component:
+
+- Produce OpenAPI dynamically based on the component configuration, e.g.
+  customize the path of controller endpoints.
+
+- Exclude certain controller endpoints based on configuration. For example, our
+  REST API explorer has a flag `useSelfHostedSpec` that controls whether a new
+  endpoint serving OpenAPI document is added.
+
+## Dynamic OpenAPI metadata
```
Comment (raymondfeng) [inline docs/site/extending/rest-api.md:43]: We could use https://loopback.io/doc/en/lb4/Booting-an-Application.html#boot-an-application-using-component for some of the use cases. I was able to modify the binding key and other metadata for services/controllers/... loaded by booters by override the `app.service()`, `app.controller()` for the subapp.

File: docs/site/migration/components/rest-api.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+alternative to low-level Express route API.
+
+If it's not feasible to implement REST endpoints as LoopBack Controllers:
+
+1. Follow the instructions in
+   [Express routes](../../extending/rest-api.md#express-routes) to setup an
+   Express router in your LoopBack 4 component and contribute it to the target
+   application.
+
+2. Take the implementation of Express routes from your LoopBack 3 component and
```
Comment (raymondfeng) [inline docs/site/migration/components/rest-api.md:106]: Express routes can be added to  a `router` using `router.use(). The `router` can be plugged into LoopBack as an ExpressRequestHandler (express middleware handler function).

File: docs/site/extending/rest-api.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+```ts
+class MyController {
+  @get('/ping')
+  ping() {
+    return {running: true};
+  }
+}
+
+export class MyComponent {
```
Comment (jannyHou) [inline docs/site/extending/rest-api.md:23]:
```suggestion
export class MyComponent implements Component {
```
Nitpick: missing `implements Component`.

File: docs/site/extending/rest-api.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+```ts
+import {config} from '@loopback/context';
+import {Component} from '@loopback/core';
+import {RestApplication} from '@loopback/rest';
+import {MyComponentBindings} from './my-component.keys.ts';
+import {definePingController} from './controllers/ping.controller-factory.ts';
+
+@bind({tags: {[ContextTags.KEY]: MyComponentBindings.COMPONENT.key}})
+export class MyComponent {
```
Comment (jannyHou) [inline docs/site/extending/rest-api.md:63]:
```suggestion
export class MyComponent implements Component {
```

File: docs/site/extending/rest-api.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+import {bind, config, ContextTags} from '@loopback/context';
+import {MyComponentBindings} from './my-component.keys.ts';
+import {PingController, StatsController} from './controllers';
+
+export interface MyComponentConfig {
+  stats: boolean;
+}
+
+@bind({tags: {[ContextTags.KEY]: MyComponentBindings.COMPONENT.key}})
+export class MyComponent {
```
Comment (jannyHou) [inline docs/site/extending/rest-api.md:111]:
```suggestion
export class MyComponent implements Component {
```

File: docs/site/extending/rest-api.md (hunk 1 above)
Comment (bajtos) [reply docs/site/extending/rest-api.md:43]: Thank you for the pointer, I agree "booting an app using component" is a concept that can be useful when migrating LB3 extensions to LB4. However, I don't see how to apply it to "Dynamic OpenAPI metadata". I opened a new issue to document "booting an app using component" for extension authors in a holistic way - see https://github.com/strongloop/loopback-next/issues/5422

File: docs/site/migration/components/rest-api.md (hunk 2 above)
Comment (bajtos) [reply docs/site/migration/components/rest-api.md:106]: > Express routes can be added to a router using `router.use()`. 

I am going to clarify this in https://github.com/strongloop/loopback-next/blob/docs/component-migration-rest-api/docs/site/extending/rest-api.md#express-routes

 > The `router` can be plugged into LoopBack as an ExpressRequestHandler (express middleware handler function).

I am not familiar with this new style. For now, I am recommending `app.mountExpressRouter()`, which has the benefit of allowing users to provide OpenAPI spec description of the Express routes too.

I opened https://github.com/strongloop/loopback-next/issues/5423 to update the docs for extension authors.

**Comment by bajtos** [comment]
Thank you @raymondfeng and @jannyHou for the review and comments, I highly appreciate them :+1:

**Pull request merged**